lib/datadog/core/normalizer.rb

```diff
@@ -0,0 +1,38 @@
+# frozen_string_literal: true
+
+module Datadog
+  module Core
+    module Normalizer
+      module_function
+
+      INVALID_TAG_CHARACTERS = %r{[^\p{L}0-9_\-:./]}
+      LEADING_INVALID_CHARS = %r{\A[^\p{L}:]++}
+      TRAILING_UNDERSCORES = %r{_++\z}
+      MAX_CHARACTER_LENGTH = 200
+
+      # Based on https://github.com/DataDog/datadog-agent/blob/45799c842bbd216bcda208737f9f11cade6fdd95/pkg/trace/traceutil/normalize.go#L131
+      # Specifically:
+      # - Must be valid UTF-8
+      # - Invalid characters are replaced with an underscore
+      # - Leading non-letter characters are removed but colons are kept
+      # - Trailing non-letter characters are removed
+      # - Trailing underscores are removed
+      # - Consecutive underscores are merged into a single underscore
+      # - Maximum length is 200 characters
+      def self.normalize(original_value)
+        normalized_value = original_value.to_s.encode('UTF-8', invalid: :replace, undef: :replace)
+        normalized_value.strip!
+        return "" if normalized_value.empty?
+
+        normalized_value.downcase!
+        normalized_value.gsub!(INVALID_TAG_CHARACTERS, '_')
+        normalized_value.sub!(LEADING_INVALID_CHARS, "")
+        normalized_value.sub!(TRAILING_UNDERSCORES, "")
+        normalized_value.squeeze!('_') if normalized_value.include?('__')
```

Comment:
Should we use `_` for invalid values? This a character that is common in file names. It will be hard to distinguish between cases where a value contains a legitimate underscore or if it matches an invalid character. 

Can we defer this normalization to the Agent? It would be nice if we could centralize this logic instead of duplicating it across SDKs

Reply:
Hmmm, for process tags, the agent is just one of the transport targets.

But, since the transport format is a tag value for traces, we have to either live with this ambiguity around `_`, or create an escape pattern for valid `_`s.

Reply:
I updated this logic a bit yesterday to follow how the Trace Agent handles tag values, which allows for values to start with a digit: https://github.com/DataDog/dd-trace-rb/pull/5033/commits/be9587d73c300d21979ebba17d420649ce6ecf95

I don't think we need to worry about having too many `_` in a string since invalid characters are also present in tag values and service names today. 

For now I think this version is consistent with the other implementations:
- .NET: https://github.com/DataDog/dd-trace-dotnet/pull/7651/files
- Java: https://github.com/DataDog/dd-trace-java/blob/5adec51856d668b7c80ff68250eaf390dafc1fa8/internal-api/src/main/java/datadog/trace/util/TraceUtils.java#L114
- Python: https://github.com/DataDog/dd-trace-py/pull/15146/files